File: app/src/main/java/com/waracle/androidtest/DataSource/DataSourceManager.java
package com.waracle.androidtest.DataSource;

import android.os.AsyncTask;

import java.lang.ref.WeakReference;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import androidx.annotation.NonNull;

public class DataSourceManager {

    @SuppressWarnings("FieldCanBeLocal")
    private long timeToCache = 1000000L;
    private ConcurrentMap<String, DataSources> dataSourceConcurrentMap = new ConcurrentHashMap<>();

    long getTimeToCache() {
        return timeToCache;
    }


    public void clearDataSources() {
        dataSourceConcurrentMap.clear();
    }

    @SuppressWarnings("unchecked")
    public void addToMap(@NonNull String url, @NonNull DataSources dataSource, @NonNull DataSources.DataListeners imageSourceListener) {
        DataSources storedDataSource = dataSourceConcurrentMap.get(url);
        if (storedDataSource == null) {
            dataSourceConcurrentMap.put(url, dataSource);
        } else {
            dataSource = storedDataSource;
        }
        dataSource.getDataSourceListeners().add(new WeakReference<>(imageSourceListener));
        final DataSources finalDataSource = dataSource;
        AsyncTask.THREAD_POOL_EXECUTOR.execute(new Runnable() {
            @Override
            public void run() {
                finalDataSource.run();
            }
        });
    }


}

File: app/src/main/java/com/waracle/androidtest/dataSource/DataSourceManager.kt
package com.waracle.androidtest.dataSource

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentHashMap

class DataSourceManager{

    private val timeToCache = 1000000L
    private val dataSourceConcurrentMap = ConcurrentHashMap<String, DataSources>()

    fun getTimeToCache(): Long {
        return timeToCache
    }

    fun addToMap(url: String, dataSource: DataSources, imageSourceListener: DataSources.DataListeners<Any>) {
        val storedDataSource = dataSourceConcurrentMap[url]
        var dataS:DataSources = dataSource
        if (storedDataSource == null) {
            dataSourceConcurrentMap[url] = dataSource
        } else {
            dataS = storedDataSource
        }
        dataS.getDataSourceListeners().add(WeakReference(imageSourceListener))
        val finalDataSource = dataS
        GlobalScope.launch {
            finalDataSource.run()
        }
    }
}